eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Imp/UtxowSpec/Invalid.hs

```diff
@@ -185,42 +214,136 @@ spec = describe "Invalid transactions" $ do
+    submitFailingTx tx [injectFailure (ValidationTagMismatch (IsValid False) PassedUnexpectedly)]
 
   it "Invalid transaction marked as valid" $ do
-    const $ pendingWith "not implemented yet"
+    let scriptHash = withSLanguage PlutusV1 (hashPlutusScript . alwaysFailsWithDatum)
+    txIn <- produceScript scriptHash
+    let tx = mkBasicTx mkBasicTxBody & bodyTxL . inputsTxBodyL .~ [txIn]
+    expectPhase2Invalid tx
 
   it "Too many execution units for tx" $ do
-    const $ pendingWith "not implemented yet"
+    let scriptHash = withSLanguage PlutusV1 (hashPlutusScript . alwaysSucceedsWithDatum)
+    txIn <- produceScript scriptHash
+    maxExUnits <- getsNES $ nesEsL . curPParamsEpochStateL . ppMaxTxExUnitsL
+    let
+      txExUnits = maxExUnits <> ExUnits 1 1
+      prp = mkSpendingPurpose (AsIx 0)
+      dat = Data $ P.I 32
+      tx =
+        mkBasicTx mkBasicTxBody
+          & bodyTxL . inputsTxBodyL .~ [txIn]
+          & witsTxL . rdmrsTxWitsL . redeemersL <>~ Map.singleton prp (dat, txExUnits)
+    submitFailingTx tx [injectFailure $ ExUnitsTooBigUTxO maxExUnits txExUnits]
 
-  it "Missing signature for collateral input" $ do
-    const $ pendingWith "not implemented yet"
+  it "Missing witness for collateral input" $ do
+    let scriptHash = withSLanguage PlutusV1 (hashPlutusScript . alwaysSucceedsWithDatum)
+    scriptInput <- produceScript scriptHash
+    (collateralHash, collateralAddr) <- freshKeyAddr
+    collateralInput <- sendCoinTo collateralAddr $ Coin 1_000_000
+    let
+      tx =
+        mkBasicTx mkBasicTxBody
+          & bodyTxL . inputsTxBodyL <>~ [scriptInput]
+          & bodyTxL . collateralInputsTxBodyL <>~ [collateralInput]
+      isOtherWitness wit = witVKeyHash wit /= asWitness collateralHash
+      dropCollateralWitness = witsTxL . addrTxWitsL %~ Set.filter isOtherWitness
+    withPostFixup (pure . dropCollateralWitness) $
+      submitFailingTx
+        tx
+        [injectFailure $ MissingVKeyWitnessesUTXOW [asWitness collateralHash]]
 
   it "Insufficient collateral" $ do
-    const $ pendingWith "not implemented yet"
+    let scriptHash = withSLanguage PlutusV1 (hashPlutusScript . alwaysSucceedsWithDatum)
+    scriptInput <- produceScript scriptHash
+    let collateral = Coin 123
+    collateralAddr <- freshKeyAddr_
+    collateralInput <- sendCoinTo collateralAddr collateral
+    let tx =
+          mkBasicTx mkBasicTxBody
+            & bodyTxL . inputsTxBodyL <>~ [scriptInput]
+            & bodyTxL . collateralInputsTxBodyL <>~ [collateralInput]
+    txFixed <- fixupTx tx
+    percentage <- getsNES $ nesEsL . curPParamsEpochStateL . ppCollateralPercentageL . to toInteger
+    let expectedCollateral = Coin . ceiling $ txFixed ^. bodyTxL . feeTxBodyL . to unCoin * percentage % 100
+    withNoFixup $
+      submitFailingTx
+        txFixed
+        [injectFailure $ InsufficientCollateral (toDeltaCoin collateral) expectedCollateral]
 
   it "Unacceptable extra redeemer" $ do
```

Comment:
Is there a point to keep the `ExtraRedeemers` test [above](https://github.com/IntersectMBO/cardano-ledger/pull/4440/files#diff-7979152785d6937f554e7498ec4d95c523133e2ed4a10a7630c330ba0b033dc0L159)  ?
They seem to be testing the same thing , but with a different redeemer purpose. Maybe we could unify them ? 
That one is running for all acceptable languages though, so we don't adopt the same strategy here, it's maybe better to keep them separate, so we have coverage for all languages.

Reply:
I wondered about that, and if I add the all-languages improvement you suggested, they would be even more similar. I left this one in because I thought perhaps a spending redeemer went through a different code path to a minting one, and the fact that in the second test I have to recalculate the purpose because the indices have been changed by the fixup function does suggest that something is a little different.

At the least, I should put the tests next to each other and change the names to reflect the different purposes being used.

Reply:
I've refactored the two `ExtraRedeemers` tests to use a common subfunction.

I also found that the original `ExtraRedeemers` test didn't work when changed to use all languages, because Babbage and Conway include some additional predicate failures that are post-Alonzo and so can't be added here. There were two other tests that have the same problem, so I've put them all in a pre-Babbage section at the end.